### PR TITLE
[TF:TRT] Fix native segment execution with int32 tensors

### PR DESCRIPTION
**Pull request by tfeher**
If the TRTEnginOp falls back to native segment execution then we have a problem with device placement of int32 tensors:
- the native segment expects these to be on the host,
- the TRTEnginOp is assigned to  GPU, therefore all of its inputs / outputs are on the GPU.

The existing workaround 8b5102cf53 sets IntsOnDevice attribute for the native segment in case we had int32 inputs. It does not handle int32 outputs the same way, becaues that would break native segment execution of CombinedNMS. To fix int32 output, we had an explicit copy inserted into the native segment 6f2656ce2e.

The existing workround did not cover all the possible cases that can produce int32 output. This PR reverts the existing workaround (6f2656ce2e and 8b5102cf53) and inserts explicit copy operations before and after the native segment is called.

Tracker: #45481

Tagging @bixia1 for review.

### CODE REVIEW

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 1]
```diff
@@ -525,26 +574,39 @@ void TRTEngineOp::ExecuteNativeSegment(OpKernelContext* ctx,
...
+    } else {
+      OP_REQUIRES_OK_ASYNC(ctx, CopyToHost(ctx, &native_inputs, i),
+                           dummy_async_helper);
+    }
   }
   VLOG(1) << "Executing native segment: " << name();
   // Increment the reference count of the async_helper by 1. When the native
-  // segment finishes execution asynchronously, we decrement the reference count
-  // of the object.
+  // segment finishes execution asynchronously, we decrement the reference
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:588]: I think we are missing synchronization right after the for loop. we need to add something like this:

if (we issue any cudaMemcpyAsync in the for loop)
      cudaStreamSynchronize(*stream);

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 2]
```diff
@@ -525,26 +574,39 @@ void TRTEngineOp::ExecuteNativeSegment(OpKernelContext* ctx,
...
+
+        for (size_t t = 0; t < native_outputs->size(); ++t) {
+          if (native_outputs->at(t).dtype() == DT_INT32) {
+            OP_REQUIRES_OK_ASYNC(ctx,
+                                 CopyToDevice(ctx, native_outputs->at(t), t),
+                                 dummy_async_helper);
+          } else {
+            ctx->set_output(t, native_outputs->at(t));
+          }
+        }
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:608]: Similar to the above, we need to add:
if (we issue any cudaMemcpyAsync in the for loop)
      cudaStreamSynchronize(*stream);

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc [hunk 3]
```diff
@@ -503,19 +503,68 @@ TRTEngineOp::TRTEngineOp(OpKernelConstruction* context)
...
+      t.NumElements() * sizeof(int32), cudaMemcpyDeviceToHost, *stream);
+  if (ret != 0) {
+    errors::Internal("Could not copy tensor for native segment input");
+  }
+  return Status::OK();
+}
+
+// Copies native_tensor, which is in host memory to ctx->output(t), which is in
+// device memory.
+Status CopyToDevice(OpKernelContext* ctx, const Tensor& native_tensor, int t) {
```
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:536]: Similar to the above, s/CopyToDevice/CopyToDeviceAsync/

**Review by tfeher — COMMENTED** [review]
Thanks @bixia1 for the review, I have fixed the issues.

**Review by bixia1 — APPROVED** [review]

**Pull request merged**